### PR TITLE
feat: wip - program area and conditions CRUD

### LINKED ISSUE #588: User Management: `program area` and `conditions` CRUD
## What needs to be done

Create server actions to create, read, update, and delete program areas.

## Why it needs to be done

To manage program areas on the front end we need to be able to manage them on the backend

## Acceptance Criteria

_List out all expected functionality/requirements_

- *only admins can perform these functions*
_On the program areas table:_
- `list_program_area`: list all program areas
- `create_program_area`: must provide a name and list of conditions to include. This cascades to `conditions` - link must be added to the program area
- `delete_program_area`: delete program area by ID; any attached conditions must remove the program area link and become unassociated conditions
- `update_program_area`: Can update name or list of conditions

_On the conditions table_
- `list_conditions`: list all conditions

### PR DESCRIPTION
**Pull request by mcmcgrath13**
# PULL REQUEST

## Summary

What changes are being proposed?

## Related Issue

Fixes #588

## Acceptance Criteria

Please copy the acceptance criteria from your ticket and paste it here for your reviewer(s)

## Additional Information

Anything else the review team should know?

## Checklist

- [ ] If this code affects the other scrum team, have they been notified? (In Slack, as reviewers, etc.)

### CODE REVIEW

File: containers/ecr-viewer/integration/programAreaService.test.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    expect(afterDelete).toBeArrayOfSize(3);
+    expect(afterCreate).toBeArrayOfSize(4);
+
+    const afterUserProgramAreas = await listUserProgramAreas(adminId!);
+    expect(afterUserProgramAreas).toBeArrayOfSize(1);
+    expect(
+      afterUserProgramAreas.filter((p) => p.uuid === progId!),
+    ).toBeArrayOfSize(1);
+  });
+});
```
Comment (gordonfarrell) [inline containers/ecr-viewer/integration/programAreaService.test.ts:153]: Nice tests!

**Review by gordonfarrell — CHANGES_REQUESTED** [review]
Overall this looks great, though my one thought is we might want a few tests around error handling to up code cov?

**Comment by mcmcgrath13** [comment]
> Overall this looks great, though my one thought is we might want a few more tests around error handling to up code cov?

@gordonfarrell I added some more coverage for some cases that are on the common-er side of corner. Some of these try/catch'es are a bit more speculative so we get more even error handling, so I'm not sure of an easy/natural way to test. Would you prefer to leave them in or remove them and let the raw errors go through?

**Comment by gordonfarrell** [comment]
> > Overall this looks great, though my one thought is we might want a few more tests around error handling to up code cov?
> 
> @gordonfarrell I added some more coverage for some cases that are on the common-er side of corner. Some of these try/catch'es are a bit more speculative so we get more even error handling, so I'm not sure of an easy/natural way to test. Would you prefer to leave them in or remove them and let the raw errors go through?

Leaving them in for now makes sense to me, it'll all get more refined as we move forward with actually implementing this. Thanks Mary!

**Review by gordonfarrell — APPROVED** [review]
Looks great!

**Review by akasper — APPROVED** [review]
Overall, the code is excellent. There are some small opportunities to expand upon the JSDoc for some of the new functions, but if anything, I think these can wait until we have a "real" implementation being used on top of these changes. 

I would also add that the test coverage is more than enough, but there is an opportunity to be obsessive about covering corner cases and/or further defining the specification:

* What happens if a program area is named `""`?
* What happens if `name` exceeds a certain length?
* Should `name`s be case-sensitive?
* What about database failures during transaction-like changes (such as updating the list of conditions)?
* [insert lengthy "yes, but what if 12 admins are making changes to a program area at the same time!?" speculations here]
* Are there any possible caching or other login-status-related errors that could make `getCheckAdmin` misbehave where it integrates with `getLoggedInUser`?

Being real: None of this is required in order for this PR to make sense. LGTM :shipit:

File: containers/ecr-viewer/src/app/services/programAreaService.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+/**
+ * Update a program with the the given uuid.
+ * @param uuid (current) id of the program area to update
+ * @param updates object with fields to update in the record.
+ * @param updates.name string of the new name for the program. Optional.
+ * @param updates.conditions list of condition codes to associate with the program (must be full list). Optional.
+ */
+export const updateProgramArea = async (
```
[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply containers/ecr-viewer/src/app/services/programAreaService.ts:56]: if there is no conditions passed, the conditions will remain whatever they were. If `[]` conditions is passed, then the program will be associated with no conditions at the end of the day

File: containers/ecr-viewer/src/app/services/userService.ts [hunk 3]
```diff
@@ -50,6 +50,21 @@ export const getLoggedInUser = cache(async () => {
 const isAdmin = (user: User | undefined): user is User =>
   !!user && user.user_type === "admin" && user.status === "active";
 
+/**
+ * Check the currently logged in user is an admin and return them. Throws
+ * an error if the currently logged in user isn't an admin.
+ * @param actionDesc description of the action that only admins can do
+ * @returns admin user
+ */
```
[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply containers/ecr-viewer/src/app/services/userService.ts:58]: We haven't been using that tag in general and most things can throw - what is the value it adds?

File: containers/ecr-viewer/src/app/services/programAreaService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcmcgrath13) [reply containers/ecr-viewer/src/app/services/programAreaService.ts:57]: clarified this in the comment
